File: Easy/0234_palindrome-linked-list/palindrome-linked-list.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	var prev *ListNode
	curr := slow
	for curr != nil {
		nxt := curr.Next
		curr.Next = prev
		prev = curr
		curr = nxt
	}

	left, right := head, prev
	for right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}
	return true
}

func buildLinkedList(values []int) *ListNode {
	dummy := ListNode{}
	current := &dummy
	for _, v := range values {
		current.Next = &ListNode{Val: v}
		current = current.Next
	}
	return dummy.Next
}

func printLinkedList(node *ListNode) {
	for node != nil {
		fmt.Printf("%d => ", node.Val)
		node = node.Next
	}
	fmt.Printf("nil\n")
}

func main() {
	tests := [][]int{
		{1,2,3,4,5},
		{1,2},
		{1},
	}

	for _, t := range tests {
		l := buildLinkedList(t)
		fmt.Printf("---\nList: ")
		printLinkedList(l)
		fmt.Printf("isPalindrome: %t\n", isPalindrome(l))
	}
}

File: Easy/0234_palindrome-linked-list/palindrome-linked-list.py
#!/bin/python3

from typing import List, Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True

        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        curr = slow
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        left, right = head, prev
        while right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        return True

def buildLinkedList(values: List[int]) -> Optional[ListNode]:
    dummy = ListNode()
    current = dummy
    for v in values:
        current.next = ListNode(v)
        current = current.next
    return dummy.next

def printLinkedList(node: Optional[ListNode]) -> None:
    while (node):
        print(node.val, end=" -> ")
        node = node.next
    print("None")

def main():
    tests: List[List[int]] = [
        [1,2,2,1],
        [1,2],
        [1]
  ]

    for t in tests:
        l = buildLinkedList(t)
        print(f"---\nList: ",end='')
        printLinkedList(l)
        print(f"isPalindrome: {Solution().isPalindrome(l)}")

if __name__ == "__main__":
    main()

